linesearch.go

```diff
@@ -50,15 +49,20 @@ func (l *Linesearch) Init(loc *Location, f *FunctionInfo, xNext []float64) (Eval
 	floats.AddScaledTo(xNext, l.initLoc, stepSize, l.direction)
 	l.funInfo = f
 	l.lastEvalType = evalType
-	l.finished = false
-	return evalType, MajorIteration, nil
+	l.nextIterType = MinorIteration
+	return evalType, l.nextIterType, nil
 }
 
 func (l *Linesearch) Iterate(loc *Location, xNext []float64) (EvaluationType, IterationType, error) {
-	if l.finished {
-		// Means that we needed to evaluate the gradient, so now we have it and can initialize
-		l.finished = false
-		loc.F = l.finishedF
+	if l.nextIterType == SubIteration {
+		// We needed to evaluate the gradient, so now we have it and can
+		// announce MajorIteration.
+		l.nextIterType = MajorIteration
```

Comment:
While true for conjugate gradient, this is wrong in the general case. For example, if the gradient is very expensive to evaluate, a linesearch may first evaluate the function. If the function value is worse, it knows it should backtrack, otherwise it then evaluates the gradient to see if it meets the armijio condition. Just doing a two-stage evaluation of the gradient does not signify a MajorIteration -- the linesearch converging is the only signifier.